checks/gcs_bucket.go

```diff
@@ -0,0 +1,130 @@
+package checks
+
+import (
+	"fmt"
+	"io/fs"
+	"time"
+
+	gcs "cloud.google.com/go/storage"
+	"github.com/flanksource/canary-checker/api/context"
+	"github.com/flanksource/canary-checker/api/external"
+	v1 "github.com/flanksource/canary-checker/api/v1"
+	"github.com/flanksource/canary-checker/pkg"
+	gcpUtils "github.com/flanksource/canary-checker/pkg/clients/gcp"
+)
+
+type GCSBucketChecker struct {
+}
+
+// Run: Check every entry from config according to Checker interface
+// Returns check result and metrics
+func (c *GCSBucketChecker) Run(ctx *context.Context) []*pkg.CheckResult {
+	var results []*pkg.CheckResult
+	for _, conf := range ctx.Canary.Spec.GCSBucket {
+		results = append(results, c.Check(ctx, conf))
+	}
+	return results
+}
+
+// Type: returns checker type
+func (c *GCSBucketChecker) Type() string {
+	return "gcsBucket"
+}
+
+type GCS struct {
+	BucketName string
+	*gcs.Client
+}
+
+type GCSFileInfo struct {
+	obj *gcs.ObjectAttrs
+}
+
+func (GCSFileInfo) IsDir() bool {
+	return false
+}
+
+func (obj GCSFileInfo) ModTime() time.Time {
+	return obj.obj.Updated
+}
+
+func (obj GCSFileInfo) Mode() fs.FileMode {
+	return fs.FileMode(0644)
+}
+
+func (obj GCSFileInfo) Name() string {
+	return obj.obj.Name
+}
+
+func (obj GCSFileInfo) Size() int64 {
+	return obj.obj.Size
+}
+
+func (obj GCSFileInfo) Sys() interface{} {
+	return obj.obj
+}
+
+func (c *GCSBucketChecker) Check(ctx *context.Context, extConfig external.Check) *pkg.CheckResult {
+	check := extConfig.(v1.GCSBucketCheck)
+	result := pkg.Success(check)
+	cfg, err := gcpUtils.NewSession(ctx, check.GCPConnection)
+	if err != nil {
+		return result
+	}
+	client := GCS{
+		BucketName: check.BucketName,
+		Client:     cfg,
+	}
+	folders, err := client.CheckFolder(ctx, check.GCSQuery)
+	if err != nil {
+		return result.ErrorMessage(err)
+	}
+	result.AddDetails(folders)
+	if test := folders.Test(check.FolderTest); test != "" {
+		result.Failf(test)
+	}
+	return result
+}
+
+func (conn *GCS) CheckFolder(ctx *context.Context, query *v1.GCSQuery) (*FolderCheck, error) {
+	result := FolderCheck{}
+	var objs *gcs.ObjectIterator
+	bucket := conn.Bucket(conn.BucketName)
+	if query != nil {
+		objs = bucket.Objects(ctx, &gcs.Query{
+			Prefix:      query.Prefix,
+			Versions:    query.Versions,
+			Delimiter:   query.Delimiter,
+			StartOffset: query.StartOffset,
+			EndOffset:   query.EndOffset,
```

Comment:
These shouldn't be specified by a user - we need to page through all the results.